### DISCUSSION TITLE
How to combine two Validation<Error, Unit>?

### DISCUSSION THREAD

**Discussion by davidyu2023** [louthy/language-ext#1292] [Q&A] [answered]
I am not sure if I am heading the right direction. If both are Success, then combining the two would return Success. Otherwise return Fail with all Errors from both. How do I do this?

**Comment by louthy** [comment, marked as answer]
Use the 'coalescing' operator:

```c#
    Validation<Error, Unit> mx = ...;
    Validation<Error, Unit> my = ...;

    var mr = mx | my;
```
I kinda wish I'd used the `&` operator, but there it is.  

You can also use `Apply` (which is the [applicative-functor](https://en.wikipedia.org/wiki/Applicative_functor) operator), it allows you to get at the values of the validation if they all succeed, and returns all the errors if any fail.

```c#
   var mr = (mx, my).Apply((x, y) => unit);
```

Finally, if you only care about getting the first error and not all errors:

```c#
    var mr = from x in mx
             from y in my
             select unit;
```

Definitely finally, there's a nice [`ValidateCreditCard` example](https://github.com/louthy/language-ext/blob/main/LanguageExt.Tests/ValidationTests.cs#L257) in the unit-tests that visits all the different approaches to using the `Validation` type, it's worth a look.

**Reply by davidyu2023** [reply]
Thanks a lot @louthy!

**Reply by louthy** [reply]
No problem @davidyu2023 and thanks for the donation, it's very much appreciated :-)

**Reply by BrunoJuchli** [reply]
@louthy 
Is there also something like an `Apply` function to be applied to an `IEnumerable<Validation<FAIL,SUCCESS>>` instead of a tuple?
So that in the end I get something like `Validation<FAIL,SEQ<SUCCESS>>`?
Well actually, I would convert the `SEQ<SUCCESS>` to another type `R`, but whether I do that inside the `Apply` or with a followup `Map` doesn't matter that much to me.

Thank you

**Reply by harrhp** [reply]
`validations.Sequence().Map(f)`
